components/director/internal/systemfetcher/client.go

```diff
@@ -0,0 +1,133 @@
+package systemfetcher
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	urlpkg "net/url"
+	"time"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/log"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+type OAuth2Config struct {
+	ClientID             string   `envconfig:"APP_OAUTH_CLIENT_ID"`
+	ClientSecret         string   `envconfig:"APP_OAUTH_CLIENT_SECRET"`
+	TokenEndpointPattern string   `envconfig:"APP_OAUTH_TOKEN_ENDPOINT_PATTERN"`
+	TenantHeaderName     string   `envconfig:"APP_OAUTH_TENANT_HEADER_NAME"`
+	ScopesClaim          []string `envconfig:"APP_OAUTH_SCOPES_CLAIM"`
+}
+
+type APIConfig struct {
+	Endpoint                    string        `envconfig:"APP_SYSTEM_INFORMATION_ENDPOINT"`
+	FilterCriteria              string        `envconfig:"APP_SYSTEM_INFORMATION_FILTER_CRITERIA"`
+	FilterTenantCriteriaPattern string        `envconfig:"APP_SYSTEM_INFORMATION_FILTER_TENANT_CRITERIA_PATTERN"`
+	Timeout                     time.Duration `envconfig:"APP_SYSTEM_INFORMATION_FETCH_TIMEOUT"`
+}
+
+type Client struct {
+	apiConfig     APIConfig
+	oAuth2Config  OAuth2Config
+	clientCreator clientCreatorFunc
+}
+
+type clientCreatorFunc func(ctx context.Context, oauth2Config OAuth2Config) *http.Client
+
+func DefaultClientCreator(ctx context.Context, oAuth2Config OAuth2Config) *http.Client {
+	cfg := clientcredentials.Config{
+		ClientID:     oAuth2Config.ClientID,
+		ClientSecret: oAuth2Config.ClientSecret,
+		TokenURL:     oAuth2Config.TokenEndpointPattern,
+		Scopes:       oAuth2Config.ScopesClaim,
+	}
+
+	httpClient := cfg.Client(ctx)
+	return httpClient
+}
+
+func NewClient(apiConfig APIConfig, oAuth2Config OAuth2Config, clientCreator clientCreatorFunc) *Client {
+	return &Client{
+		apiConfig:     apiConfig,
+		oAuth2Config:  oAuth2Config,
+		clientCreator: clientCreator,
+	}
+}
+
+// FetchSystemsForTenant fetches systems from the service by making 2 HTTP calls with different filter criteria
+func (c *Client) FetchSystemsForTenant(ctx context.Context, tenant string) ([]System, error) {
+	ctx, httpClient := c.clientForTenant(ctx, tenant)
```

Comment:
We can think of a way to avoid creating a new HTTP client for each tenant to avoid using too much unnecessary memory. Maybe we can use a single HTTP client here in the SystemFetcher `client.go` and we can use `clientcredentials.Config{...}.Token(...)` to fetch only tokens for each tenant and just update the `Authorization` header of the SystemFetcher's HTTP client.